rpc/client.go

```diff
@@ -118,53 +116,53 @@ func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
 }
 
 // connect dials the connection in a backoff/retry loop.
-func (c *Client) connect(opts *retry.Options, context *Context) {
+func (c *Client) connect(opts *retry.Config, context *Context) {
 	// Attempt to dial connection.
 	retryOpts := clientRetryOptions
 	if opts != nil {
 		retryOpts = *opts
 	}
-	retryOpts.Tag = fmt.Sprintf("client %s connection", c.addr)
 	retryOpts.Stopper = context.Stopper
 
-	err := retry.WithBackoff(retryOpts, func() (retry.Status, error) {
-		tlsConfig, err := context.GetClientTLSConfig()
-		if err != nil {
-			// Problem loading the TLS config. Retrying will not help.
-			return retry.Break, err
-		}
-
-		conn, err := tlsDialHTTP(c.addr.Network(), c.addr.String(), tlsConfig)
-		if err != nil {
-			// Retry if the error is temporary, otherwise fail fast.
-			if t, ok := err.(net.Error); ok && t.Temporary() {
-				return retry.Continue, err
+	if err := func() error {
+		for r := retry.Start(retryOpts); r.Next(); {
+			tlsConfig, err := context.GetClientTLSConfig()
+			if err != nil {
+				// Problem loading the TLS config. Retrying will not help.
+				return err
 			}
-			return retry.Break, err
-		}
 
-		c.mu.Lock()
-		c.Client = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
-		c.lAddr = conn.LocalAddr()
-		c.mu.Unlock()
+			conn, err := tlsDialHTTP(c.addr.Network(), c.addr.String(), tlsConfig)
+			if err != nil {
+				// Retry if the error is temporary, otherwise fail fast.
+				if t, ok := err.(net.Error); ok && t.Temporary() {
+					continue
+				}
+				return err
+			}
 
-		// Ensure at least one heartbeat succeeds before exiting the
-		// retry loop. If it fails, don't retry: The node is probably
-		// dead.
-		if err = c.heartbeat(); err != nil {
-			return retry.Break, err
-		}
+			c.mu.Lock()
+			c.Client = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
+			c.lAddr = conn.LocalAddr()
+			c.mu.Unlock()
 
-		// Signal client is ready by closing Ready channel.
-		log.Infof("client %s connected", c.addr)
-		close(c.Ready)
+			// Ensure at least one heartbeat succeeds before exiting the
+			// retry loop. If it fails, don't retry: The node is probably
+			// dead.
+			if err = c.heartbeat(); err != nil {
+				return err
+			}
 
-		// Launch periodic heartbeat.
-		go c.startHeartbeat()
+			// Signal client is ready by closing Ready channel.
+			log.Infof("client %s connected", c.addr)
+			close(c.Ready)
 
-		return retry.Break, nil
-	})
-	if err != nil {
+			// Launch periodic heartbeat.
+			go c.startHeartbeat()
+			return nil
+		}
+		return nil
+	}(); err != nil {
```

Comment:
since nothing really happens after the retry loop, you could `defer` the optional log & close before, and have `connect()` return an error (which needs to be named for the defer). That would save the awkward `func` in exchange for a named parameter. If you go for it, might make sense to also move the log to the callsite (there's only one).